curp/src/server/spec_pool.rs

```diff
@@ -1,61 +1,52 @@
-use std::collections::{HashMap, VecDeque};
+use std::sync::Arc;
 
-use tokio::time::Instant;
+use indexmap::IndexMap;
+use parking_lot::Mutex;
 use tracing::debug;
 
 use crate::cmd::{Command, ProposeId};
 
+/// A reference to the speculative pool
+pub(super) type SpecPoolRef<C> = Arc<Mutex<SpeculativePool<C>>>;
+
 /// The speculative pool that stores commands that might be executed speculatively
 #[derive(Debug)]
 pub(super) struct SpeculativePool<C> {
     /// Store
-    pub(super) pool: VecDeque<C>,
-    /// Store the ids of commands that have completed backend syncing, but not in the local speculative pool. It'll prevent the late arrived commands.
-    pub(super) ready: HashMap<ProposeId, Instant>,
+    pub(super) pool: IndexMap<ProposeId, Arc<C>>,
 }
 
 impl<C: Command + 'static> SpeculativePool<C> {
     /// Create a new speculative pool
     pub(super) fn new() -> Self {
         Self {
-            pool: VecDeque::new(),
-            ready: HashMap::new(),
+            pool: IndexMap::new(),
         }
     }
 
-    /// Push a new command into spec pool if it has not been marked ready
-    pub(super) fn push(&mut self, cmd: C) {
-        if self.ready.remove(cmd.id()).is_none() {
+    /// Push a new command into spec pool if it has no conflict
+    pub(super) fn try_insert(&mut self, cmd: Arc<C>) -> bool {
```

Comment:
We can directly rename this method as `insert`. `try_insert` looks like it will return a `Result`

Reply:
Sure